### PR TITLE
TF2 porting: h3 feature (re-submission)

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Code Pull Requests

This is still work-in-progress.  I wanted to give you an early look into the progress.  Right now I have the `embed` and `weighted_sum` passing the unit test.  I have not started on `rnn` encoder yet.

```
pytest -v test_experiment.py::test_experiment_h3
=========================================================== test session starts ============================================================
platform linux -- Python 3.6.9, pytest-5.4.3, py-1.8.1, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /opt/project
plugins: pycharm-0.6.0, typeguard-2.8.0
collected 3 items

test_experiment.py::test_experiment_h3[embed] PASSED                                                                                 [ 33%]
test_experiment.py::test_experiment_h3[weighted_sum] PASSED                                                                          [ 66%]
test_experiment.py::test_experiment_h3[rnn] FAILED                                                                                   [100%]
```

re: the `weighted_sum` encoder, I had to rename the `self.weights` instance variable.  This came about when I made `H3WeightedSum` a subclass of `tf.keras.layers.Layer`.  It appears that `tf.keras.layers.Layer` has a read-only attribute called `weights`.    This caused an error when the `H3WeightedSum` constructor executed.  

I renamed `self.weights` to `self.h3_weights` and modified the subsequent code as needed.  AFAICT, I made all adjustments for the new name of `self.h3_weights`.  I'd appreciate if you double check this change.

Second to better understand the H3 encoder, I reviewed the ludwig user doc.  I noticed a issue.  The description of the H3 encoder is word for word the same as for the Date encoder:
```
H3 Input Features and Encoders¶

Input date features are transformed into a int valued tensors of size N x 8 (where N is the size
of the dataset and the 8 dimensions contain year, month, day, weekday, yearday, hour, 
minute and second) and added to HDF5 with a key that reflects the name of column in the CSV.

Currently there are two encoders supported for dates: Embed Encoder and Wave encoder 
which can be set by setting encoder parameter to embed or wave in the input feature dictionary 
in the model definition (embed is the default one).
```
I was going to open an issue re: this observation but I noticed all the documentation content was moved out of the repo a few days ago.  So I'm not sure where to open the issue.

### CODE REVIEW

**Comment by jimthompson5802** [comment]
The last commit finishes the work.  All three unit tests pass for the h3 feature
```
pytest -v test_experiment.py::test_experiment_h3
=========================================================== test session starts ============================================================
platform linux -- Python 3.6.9, pytest-5.4.3, py-1.8.1, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /opt/project
plugins: pycharm-0.6.0, typeguard-2.8.0
collected 3 items

test_experiment.py::test_experiment_h3[embed] PASSED                                                                                 [ 33%]
test_experiment.py::test_experiment_h3[weighted_sum] PASSED                                                                          [ 66%]
test_experiment.py::test_experiment_h3[rnn] PASSED                                                                                   [100%]

============================================================= warnings summary =============================================================
/usr/local/lib/python3.6/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py:15
  /usr/local/lib/python3.6/dist-packages/tensorflow/python/pywrap_tensorflow_internal.py:15: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/integration_tests/test_experiment.py::test_experiment_h3[embed]
tests/integration_tests/test_experiment.py::test_experiment_h3[embed]
tests/integration_tests/test_experiment.py::test_experiment_h3[weighted_sum]
tests/integration_tests/test_experiment.py::test_experiment_h3[weighted_sum]
tests/integration_tests/test_experiment.py::test_experiment_h3[rnn]
tests/integration_tests/test_experiment.py::test_experiment_h3[rnn]
  /usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:71: FutureWarning: Pass pos_label=None, sample_weight=None as keyword args. From version 0.25 passing these as positional arguments will result in an error
    FutureWarning)

tests/integration_tests/test_experiment.py::test_experiment_h3[embed]
tests/integration_tests/test_experiment.py::test_experiment_h3[weighted_sum]
tests/integration_tests/test_experiment.py::test_experiment_h3[rnn]
  /usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:71: FutureWarning: Pass average=macro, sample_weight=None as keyword args. From version 0.25 passing these as positional arguments will result in an error
    FutureWarning)

tests/integration_tests/test_experiment.py::test_experiment_h3[embed]
tests/integration_tests/test_experiment.py::test_experiment_h3[weighted_sum]
tests/integration_tests/test_experiment.py::test_experiment_h3[rnn]
  /usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:71: FutureWarning: Pass average=micro, sample_weight=None as keyword args. From version 0.25 passing these as positional arguments will result in an error
    FutureWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
====================================================== 3 passed, 13 warnings in 4.02s ====================================================
```

**Comment by w4nderlust** [comment]
> re: the `weighted_sum` encoder, I had to rename the `self.weights` instance variable. This came about when I made `H3WeightedSum` a subclass of `tf.keras.layers.Layer`. It appears that `tf.keras.layers.Layer` has a read-only attribute called `weights`. This caused an error when the `H3WeightedSum` constructor executed.
> 
> I renamed `self.weights` to `self.h3_weights` and modified the subsequent code as needed. AFAICT, I made all adjustments for the new name of `self.h3_weights`. I'd appreciate if you double check this change.

I would call it `self.aggregation_weights` to be more descriptinve, but ather than that it looks fine to me!

> Second to better understand the H3 encoder, I reviewed the ludwig user doc. I noticed a issue. The description of the H3 encoder is word for word the same as for the Date encoder:
> 
> ```
> H3 Input Features and Encoders¶
> 
> Input date features are transformed into a int valued tensors of size N x 8 (where N is the size
> of the dataset and the 8 dimensions contain year, month, day, weekday, yearday, hour, 
> minute and second) and added to HDF5 with a key that reflects the name of column in the CSV.
> 
> Currently there are two encoders supported for dates: Embed Encoder and Wave encoder 
> which can be set by setting encoder parameter to embed or wave in the input feature dictionary 
> in the model definition (embed is the default one).
> ```
> 
> I was going to open an issue re: this observation but I noticed all the documentation content was moved out of the repo a few days ago. So I'm not sure where to open the issue.

mmm it must have been some copy-paste errors in the past. I will rewrite that part and maybe dig into the history to see if I can find where theoriginal description went.
Anyway the docs are here: https://github.com/ludwig-ai/ludwig-doc
> The last commit finishes the work. All three unit tests pass for the h3 feature

That's great, let me do a pass, in particular on the encoders that are the most tricky thing about this PR.

File: ludwig/features/h3_feature.py [hunk 1]
```diff
@@ -77,56 +77,30 @@ def add_feature_data(
...
 
-    def get_h3_encoder(self, encoder_parameters):
-        return get_from_registry(
-            self.encoder, h3_encoder_registry)(
-            **encoder_parameters
-        )
+    def call(self, inputs, training=None, mask=None):
+        assert isinstance(inputs, tf.Tensor)
+        # assert inputs.dtype == tf.float32 or inputs.dtype == tf.float64
+        # assert len(inputs.shape) == 1
```
Comment (w4nderlust) [inline ludwig/features/h3_feature.py:96]: This should be == 2

File: ludwig/features/h3_feature.py [hunk 2]
```diff
@@ -77,56 +77,30 @@ def add_feature_data(
...
-        self.encoder_obj = self.get_h3_encoder(encoder_parameters)
 
-    def get_h3_encoder(self, encoder_parameters):
-        return get_from_registry(
-            self.encoder, h3_encoder_registry)(
-            **encoder_parameters
-        )
+    def call(self, inputs, training=None, mask=None):
+        assert isinstance(inputs, tf.Tensor)
+        # assert inputs.dtype == tf.float32 or inputs.dtype == tf.float64
```
Comment (w4nderlust) [inline ludwig/features/h3_feature.py:95]: This should be integers, specifically uint8, but I'm not 100% sure of the casting rules here. I think uint8 can be casted to int32 because of embeddings_lookup needing it for selecting the rows of the embedding matrix, but inputs should arrive here as uint8

File: ludwig/models/modules/h3_encoders.py [hunk 3]
```diff
@@ -330,13 +319,13 @@ def __init__(
             dropout=dropout_rate,
             initializer=weights_initializer,
             regularize=weights_regularizer,
-            reduce_output=None,
+            reduce_output='None',
```
Comment (w4nderlust) [inline ludwig/models/modules/h3_encoders.py:322]: I believe `None` should work too, doesn't it?

File: ludwig/features/h3_feature.py (hunk 1 above)
Comment (jimthompson5802) [reply ludwig/features/h3_feature.py:96]: made change

File: ludwig/features/h3_feature.py (hunk 2 above)
Comment (jimthompson5802) [reply ludwig/features/h3_feature.py:95]: made the change

File: ludwig/models/modules/h3_encoders.py [hunk 4]
```diff
@@ -495,7 +477,7 @@ def __init__(
             dropout=dropout,
             initializer=initializer,
             regularize=regularize,
-            reduce_output=None
+            reduce_output='None'
         )
 
         self.recurrent_stack = RecurrentStack(
```
Comment (w4nderlust) [inline ludwig/models/modules/h3_encoders.py:483]: here if the recurrent stack has additional parameteres that are not surfaced we should probably add them to the init of h3rnn. If those are all the parameters, then we are good.

File: ludwig/models/modules/h3_encoders.py [hunk 5]
```diff
@@ -508,40 +490,38 @@ def __init__(
...
-            :param is_training: Tesnor (tf.bool) specifying if in training mode
-                   (important for dropout)
-            :type is_training: Tensor
-        """
+            :param training: bool specifying if in training mode (important for dropout)
+            :type training: bool
+            :param mask: bool tensor encoding masked timesteps in the input
+            :type mask: bool
+         """
+        input_vector = inputs
```
Comment (w4nderlust) [inline ludwig/models/modules/h3_encoders.py:508]: Probably this renaming is not needed

File: tests/integration_tests/test_experiment.py [hunk 6]
```diff
@@ -613,16 +612,16 @@ def test_experiment_date(csv_filename):
         input_features[0]['encoder'] = encoder
         run_experiment(input_features, output_features, data_csv=rel_path)
 
-
-def test_experiment_h3(csv_filename):
+@pytest.mark.parametrize('encoder', ['embed', 'weighted_sum', 'rnn'])
```
Comment (w4nderlust) [inline tests/integration_tests/test_experiment.py:615]: Maybe here we can use `H3InputFeature.encoder_registry.keys()` instad, so that is we add a new one we are sure it is tested.

File: ludwig/models/modules/h3_encoders.py (hunk 3 above)
Comment (jimthompson5802) [reply ludwig/models/modules/h3_encoders.py:322]: Yeah it works with `None` (not as a string)...the reason I changed it was PyCharm was flagging it a warning.  Here is the warning I saw
Screen Shot 2020-06-12 at 20 42 28
I can put it back to the `None` type

File: ludwig/models/modules/h3_encoders.py (hunk 5 above)
Comment (jimthompson5802) [reply ludwig/models/modules/h3_encoders.py:508]: removed the rename

File: ludwig/models/modules/h3_encoders.py (hunk 3 above)
Comment (w4nderlust) [reply ludwig/models/modules/h3_encoders.py:322]: It uses `reduce_sequence` behind the sceens, which work with "none", "None" and None. But it's ok to use the string too, maybe in the future I can add more type annotations to avoid these wrong type inferences from the ide.

File: tests/integration_tests/test_experiment.py (hunk 6 above)
Comment (jimthompson5802) [reply tests/integration_tests/test_experiment.py:615]: changed to use `H3InputFeature.encoder_registry.keys()`

File: ludwig/models/modules/h3_encoders.py (hunk 4 above)
Comment (jimthompson5802) [reply ludwig/models/modules/h3_encoders.py:483]: While working to surface the recurrent stack parameters, I noticed this
https://github.com/jimthompson5802/ludwig/blob/8a069c1ce8e8994825999c5de5ece478d70300d2/ludwig/models/modules/h3_encoders.py#L392-L406.  Line 402 has `droput=False`.  From what I can tell, this should be a float between 0.0 and 1.0.  In theory, `False` numerically evaluates to 0.0. I did a quick skim how `dropout` is used.  From what I can tell, it is used only as a numerical attribute.  

Any issue with changing from `dropout=False` to `dropout=0.0`?

Comment (w4nderlust) [reply ludwig/models/modules/h3_encoders.py:483]: Yes it should be changed to `dropout_rate` like everywhere else, and it should be 0.0 by default. Also, the surfacing thing should be done for the FClayers too, and also in the date feature is needed. Didn't realize it when you sbmitted the date PR, but in general the principle is: surface as many parameters from the layers used inside the encoder as possible (or as makes sense). For the sequence feature encoders for instance I spent quite some time to make sure all the relevant parameters are surfaced, we should make sure the same happens in all other features.

Comment (jimthompson5802) [reply ludwig/models/modules/h3_encoders.py:483]: added remaining recurrent stack parameters, changed `dropout` to `dropout_rate`.

**Comment by jimthompson5802** [comment]
The last set of commits addressed all your comments.  

If this looks good, I'll go back to Date encoder to make sure about surfacing the internal layer parameters.

**Comment by w4nderlust** [comment]
> The last set of commits addressed all your comments.
> 
> If this looks good, I'll go back to Date encoder to make sure about surfacing the internal layer parameters.

Looks good to me, merging now. I was thinking about surfacing also the embedding parameters, but probably there's no big need for that, so we are good to move forward.
